### DISCUSSION TITLE
How do I best handle interpolating movements of renderable objects?

### DISCUSSION THREAD

**Discussion by Architector4** [bevyengine/bevy#8923] [Q&A]
Let's say that I have a system (a physics engine or whatnot else) that updates Transform component's values of certain entities once every fixed timestep moment (for example 10 times a second, way lower than the usual framerate), and I want to represent those entities in the game's visible space, with meshes for example.

Obviously, doing it the most naive way of slapping `PbrBundle` on such an object will make it look choppy, as with the entity moves the mesh, 10 times a second. I want to render it in a way that looks smooth, somehow interpolate between the different Transform values for each rendered frame on the window, but in a way that will not disturb the original system at all, ensuring that every time it runs the Transform value is exactly the same as what it left it at the last time.

Also, I intend to use multiple systems that would run at a single fixed timestep at once, not just one.

Any suggestions on how to best approach this? My first three thoughts are:
* Try to just mash all components into one entity naively and then interpolate the Transform around in the frames when the system is not run, making it arrive exactly to where it was last left off right before the system runs again. Gotta not screw that one up.
* Run a separate ECS world that has just the entities and systems running at a fixed timestep in a controlled manner, and then sync it to the main ECS world which will have a mirroring set of Entities with renderables (`PbrBundle` and stuff) and systems which have just the interpolation logic in them. Sounds a bit cumbersome, but I think it has a nifty bonus: it sounds like it'd be easier to store the data the systems change in separate components, then run all the systems again on the spot, and get one frame of lookahead data that can be useful to perform better interpolation, and then use the stored data to revert it back.
* Try to do the above but in a singular ECS world, with both an interpolated renderable and the actual entity existing in there at once. Sounds like potentially easier to debug, but also more annoying in terms of separating behavior.

---

This post has been a bit of a braindump and I guess it might not make much sense at times. Thoughts on any of this?

**Comment by helvieq499** [comment]
The way that I would naively do it would be to have a new component that represents the real, uninterpolated position that game logic interacts with directly, and a system that runs after all game logic but before rendering that interpolates the custom component's values onto the real Transform component every frame.

I think this would work well because systems that want interpolation can opt into it, the system can query for entities with the component instead of iterating the world manually, and no extra logic needs to be added to any system that wants interpolation.

**Reply by helvieq499** [reply]
I decided to implement what I described in a crate named `bevy_mod_interp` if you would like to use it. 
It only smooths translations for now and the rate of smoothing cannot be configured, but it is framerate independent.

**Reply by Architector4** [reply]
This looks neat, but it appears that one has to have the systems use the new `InterpTransform` component instead of the usual `Transform` one. I'm more looking on coming up with something that would work with any generic system relying on hardcoded `Transform` component from Bevy's crate, for example the `bevy_rapier3d` crate.

**Reply by helvieq499** [reply]
I couldn't find a way to do it right now, but it might be possible to use change detection to set a target transformation which is interpolated to. 
The issue I encountered was that it appears that for a transform to visually move it needs to be marked as changed, but then I cannot distinguish changes made by the interpolation system and other code.
